Add Debian 12 Compatibility & Automated Installer

tiamoesg opened the pull request:
### Ticket
[Link to Github Issue
](https://github.com/tenstorrent/tt-metal/issues/18297)

### Problem description
This bounty involves enabling Tenstorrent's TT-Metal and related system engineering tools to be packaged and buildable on specific Linux distributions ([debian](https://www.debian.org/) for this bounty) that are currently unsupported. The goal is to ensure that TT-Metal and its dependencies can be cleanly compiled and installed via standard packaging methods on the targeted distributions.

### What's changed
This PR introduces full Debian 12 compatibility for the `tt-metal` stack, including a new top-level installer (`install_tt-debian.sh`) and updates to core build scripts. These changes enable a smooth, fully automated installation on fresh Debian systems, including kernel headers, swapfile, Rust, drivers, and firmware setup.

### Key Changes
- **build_metal.sh**
  - Debian RPATH/linker fixes for `wandb-cpp`.
  - Updated CMake args for LD_LIBRARY_PATH handling.
- **install_dependencies.sh**
  - Debian-specific dependency resolution.
  - Unified install flow for Ubuntu & Debian.
- **create_venv.sh**
  - Improved sequencing & virtualenv setup.
- **install_tt-debian.sh** *(new)*
  - One-command automated Debian install.
  - Configures hugepages, drivers, tt-smi, firmware, etc.
- **scripts/build_scripts/create_stubs.sh**
  - Error messaging & formatting refinements.

### Validation
- Fully tested on DigitalOcean Debian 12 droplet (4GB RAM + 8GB swap).
- Verified firmware flashing, hugepages, and `tt-smi`.

### Key Considerations:
On fresh install of Debian 12, user must update kernel and headers:
```
apt install linux-image-amd64 linux-headers-amd64
sudo reboot
```
then run the installer_tt-debian.sh script

tiamoesg committed:
Create install_tt-debian.sh

Automated installer for TT on Debian 12

tiamoesg committed:
Update install_dependencies.sh

Added Debian 12 compatibility

tiamoesg committed:
Update build_metal.sh

Added Debian 12 compatibility

tiamoesg committed:
Update create_venv.sh

Added formatting and exports for environment variables. Including device selection

tiamoesg committed:
Update create_stubs.sh

Added formatting to match install scripts

nsextonTT reviewed (changes requested):
It appears you have a bad merge or rebase in `install_dependencies.sh` and a lot of changes from `main` are overridden. Could you please fix?

tiamoesg committed:
Update install_dependencies.sh

Rebased to newest install_dependencies.sh; added debian fixes

tiamoesg commented:
> There are several team members on travel this week, so it might be a little bit. Pings have gone out.

Cool, thank you!

tiamoesg committed:
Update install_tt-debian.sh

replaced curl commands that went to my custom repo scripts. -- now that the build_metal.sh and install_dependencies.sh will be proper; they aren't needed.

tiamoesg commented on install_tt-debian.sh:
```diff
@@ -0,0 +1,160 @@
+# STEP 6: Create Python Virtual Environment
+# ----------------------------
+echo -e "${STEP} Fetching custom create_venv.sh..."
+curl -sSL https://raw.githubusercontent.com/tiamoesg/tt-metal/main/create_venv.sh -o create_venv.sh
```
forgot to take these out, they were for bypassing the tenstorrent scripts while testing.

tiamoesg commented on install_tt-debian.sh:
```diff
@@ -0,0 +1,160 @@
+# STEP 7: Build TT-Metal
+# ----------------------------
+echo -e "${STEP} Fetching custom build_metal.sh..."
+curl -sSL https://raw.githubusercontent.com/tiamoesg/tt-metal/main/build_metal.sh -o build_metal.sh
```
forgot to take these out, they were for bypassing the tenstorrent scripts while testing.

tiamoesg committed:
Update create_venv.sh

deleted duplicate line and declared red

tiamoesg committed:
Update create_venv.sh

Color definitions updated

tiamoesg commented on create_venv.sh:
```diff
@@ -1,49 +1,139 @@
+        rustup default stable
+    fi
+else
+    echo -e "${RED}Warning: rustup environment not found. Rust may be outdated.${RESET}"
```
fixed.

tiamoesg commented on create_venv.sh:
```diff
@@ -1,49 +1,139 @@
+            ;;
+    esac
+done
+echo -e "${GREEN}Selected card model: ${BOLD}$ARCH_NAME${RESET}"
```
fixed.

tiamoesg commented on create_venv.sh:
```diff
@@ -1,49 +1,139 @@
+
+echo -e "${CYAN}Installing tt-metal${RESET}"
+$PYTHON_ENV_DIR/bin/pip install -e .
+
```
fixed.

blozano-tt commented on create_venv.sh:
```diff
@@ -1,49 +1,139 @@
+RED="\e[31m"
+BLUE="\e[34m"
+BOLD="\e[1m"
+RESET="\e[0m"
```
Why not do something like this, to avoid in-band colorization?

```
log_info()    { echo -e "${CYAN}[INFO] $*${RESET}"; }
log_warn()    { echo -e "${YELLOW}[WARN] $*${RESET}"; }
log_error()   { echo -e "${RED}[ERROR] $*${RESET}"; }
```

tiamoesg replied:
Implemented your suggestion, and will be implementing the same for the install_tt-debian.sh and /scripts/build_scripts/create_stubs.sh

tiamoesg committed:
Update create_venv.sh

Added logging functions to avoid in-band colorization.

Fixed "pre-commit install" to use python_venv, to not require"source"ing the script.

tiamoesg committed:
Update install_tt-debian.sh

Added logging functions to avoid in-band colorization

tiamoesg committed:
Update create_stubs.sh

Added logging functions to avoid in-band colorization

warthog9 commented:
If it's of any help, we don't / can't support Debian 12 from the low level tooling, specifically luwen (and Debian 12 being old enough trying to sort that turned into a mess).

Do plan to support 13, so might want to shift slightly in that direction from a general packaging perspective

(note: I haven't reviewed anything yet, mostly making general note)

tiamoesg commented:
> If it's of any help, we don't / can't support Debian 12 from the low level tooling, specifically luwen (and Debian 12 being old enough trying to sort that turned into a mess).
> 
> Do plan to support 13, so might want to shift slightly in that direction from a general packaging perspective
> 
> (note: I haven't reviewed anything yet, mostly making general note)

Thanks for highlighting that! I ran into some quirks early on compiling and packaging Pyluwen, but removing rustc and installing Rust via rustup allowed the current wheels to build and work without having to package manually. I’ve added this rustup step to my install_tt-debian.sh script as part of preparing Debian 12.

My Wormhole card is running computational examples and everything in tt-metal appears to be functioning correctly. So, at least from a practical standpoint, luwen seems fully operational. (Note: I haven’t dug deeply into luwen’s internals or Rust’s role here, so I can’t 100% vouch for the lowest-level tooling.)

I did notice that multiple scripts pull in different versions of the Pyluwen wheel (7.2–7.6, if I remember correctly). They all worked fine, so I left that as-is since harmonizing versions felt out of scope for this bounty.

Looking forward to your review—thanks again!

tiamoesg commented on install_tt-debian.sh:
```diff
@@ -0,0 +1,162 @@
+# STEP 3: Install Rust (latest via rustup)
+# ----------------------------
+log_step "Installing Rust..."
+sudo apt remove -y rustc cargo || true
```
The swap file portion can be eliminated; as it is only for systems that don’t meet Tenstorrent’s minimum hardware requirements.

As for the Rust component; it currently updates the system Rust to the newest stable version. —For a dedicated system running Tenstorrent hardware and workloads, it seems to be well implemented with Rust’s current 1.88 (if I remember the version correctly). So I’d imagine issues would only be with fringe cases from unique hardcoded dependencies or legacy workflows. 

Thus in practicality; anyone can fire up a Debian system and run Tenstorrent hardware with this. Which seems pretty powerful for the community, so that may be the appropriate approach with the community folder! 

Thank you for the update; looking forward to see what feedback you get.

nsextonTT reviewed (approved):
Working on bringing this in and marking bounty as complete.

tiamoesg commented:
> I can't say for sure since the idea is still young. We'll continue to monitor and gauge community response. I'm hoping that it will be embraced and give people more reasons to give TT a try :)
> 
> I've given myself a deadline of tomorrow to get this merged in along with several other of the distro enablement bounties. Just doing a little bit of bookkeeping on my end to put a bow on it.

Awesome! Nice work and great communication along the way! -- If there is any team responsible for Community, or Bounties related to, or really anything community-building oriented from the TT side, I'd enjoy contributing towards building that up and making it more friendly for people to get involved. I've noticed some areas where strategic content creation could help bridge that gap and potentially entice people to learn more about graph processing and it's wonders. 

Thanks again for pushing on the Community folder! I'm sure it'll be a great addition. 

It's been a pleasure contributing with you, enjoy your next one!
